actions/profile_db.py

```diff
@@ -0,0 +1,435 @@
+                .filter(Transaction.to_account_number == to_account_number)
+            )
+        else:
+            transactions = self.session.query(Transaction).filter(
+                Transaction.from_account_number == account_number
+            )
+        if start_time:
+            transactions = transactions.filter(Transaction.timestamp >= start_time)
+        if end_time:
+            transactions = transactions.filter(Transaction.timestamp <= end_time)
+
+        return transactions
+
+    def list_credit_cards(self, session_id):
+        account = self.get_account_from_session_id(session_id)
+        cards = (
+            self.session.query(CreditCard)
+            .filter(CreditCard.account_id == account.id)
+            .all()
+        )
+        return [card.credit_card_name for card in cards]
+
+    def get_credit_card(self, session_id, credit_card_name):
+        account = self.get_account_from_session_id(session_id)
+        return (
+            self.session.query(CreditCard)
+            .filter(CreditCard.account_id == account.id)
+            .filter(CreditCard.credit_card_name == credit_card_name.lower())
+            .first()
+        )
+
+    def get_credit_card_balance(
+        self, session_id, credit_card_name, balance_type="current_balance"
+    ):
+        balance_type = "_".join(balance_type.split())
+        card = self.get_credit_card(session_id, credit_card_name)
+        return getattr(card, balance_type)
+
+    @staticmethod
+    def list_balance_types():
+        return [
+            " ".join(name.split("_"))
+            for name in CreditCard.__table__.columns.keys()
+            if name.endswith("balance")
+        ]
+
+    def list_vendors(self):
+        vendors = (
+            self.session.query(Account.account_holder_name)
+            .filter(Account.session_id.startswith("vendor_"))
+            .all()
+        )
+        return [vendor.account_holder_name for vendor in vendors]
+
+    def pay_off_credit_card(self, session_id, credit_card_name, amount):
+        account = self.get_account_from_session_id(session_id)
+        account_number = self.get_account_number(account)
+        credit_card = (
+            self.session.query(CreditCard)
+            .filter(CreditCard.account_id == account.id)
+            .filter(CreditCard.credit_card_name == credit_card_name.lower())
+            .first()
+        )
+        self.transact(
+            account_number, self.get_account_number(credit_card), amount,
+        )
+        credit_card.current_balance -= amount
+        if amount < credit_card.minimum_balance:
+            credit_card.minimum_balance -= amount
+        else:
+            credit_card.minimum_balance = 0
+        self.session.commit()
+
+    def add_session_account(self, session_id, name=""):
+        self.session.add(
+            Account(session_id=session_id, account_holder_name=name, currency="$")
+        )
+
+    def add_credit_cards(self, session_id):
```

Comment:
When running subsequent sessions, the table of credit cards contains duplicates:

![image](https://user-images.githubusercontent.com/16159200/105042056-caf6c780-5a31-11eb-92ee-9c1fa5cbcda7.png)

Reply:
@melindaloubser1 ,
I now understand what it is doing...
- Each session becomes a new account
- Each account then has the credit cards

So, these are not actually duplicates.

I will consider this as resolved, since it was just a misunderstanding from my side.